frontend/lib/util/filtering.cpp

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2021-2023 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chpl/util/filtering.h"
+#include <regex>
+
+namespace chpl {
+
+std::string removeSphinxMarkup(const std::string& msg) {
```

Comment:
I feel the regexp is getting a little unwieldly to read.  I wonder if by
splitting the regexp apart and adding some comments we can make things a little
easier to understand.

I'm thinking something like this:

```cpp
namespace {
  std::string buildSphinxMarkupRegexp() {
    // See documentation on inline Sphinx markup here:
    // https://chapel-lang.org/docs/latest/tools/chpldoc/chpldoc.html#inline-markup-2
    // There are some details involved in this, but the main idea is to match
    // strings of the form:
    // 
    //    :role:`content`
    // 
    // We write the regexp so the relevant part of "content" will be captured by
    // std::smatch if we wish to use this to filter out the markup.

    // TODO: Support explicit title and reference targets like in reST direct hyperlinks (and having only target
    //       show up in sanitized message).

    // There are three "kinds" of content we consider, so we OR these together
    auto reCntType1 = R"#(([$\w\$\.]+))#"         // Capture a normal Chapel identifier
    auto reCntType2 = R"#((?:~([$\w\$]+\.?)+))#"; // If it starts with ~, capture last identifier to the right of a '.'  Note in a repeated
                                                  //   capture group (e.g. `(\w\.?)+` only the last iteration is captured
    auto reCntType3 = R"#((?:!([$\w\$\.]+)))#";   // Starts with !, capture identifier to right of ! without capturing ! itself
    auto reContent =  std::format(R"#((?:{}|{}|{}))#", reCntType1, reCntType2, reCntType3);

    // Various roles; put into a (?...) group so we don't "capture" it
    auto reRole = R"#((?:mod|proc|iter|data|const|var|param|type|class|record|attr|enum))#"

    // Regexp to match :role:`content`; note \B is used on either end to indicate markup is separted by word boundaries
    auto reSphinxMrkp = std::format( R"#(\B\:{}\:`{}`\B#)", reRole, reTarget);

    return reSphinxMrkp;
  }
}

std::string removeSphinxMarkup(const std::string& msg) {
  static const auto reStr = buildSphinxMarkupRegexp();

  // ... your existing match and replace code here ...
}

```

You get the idea, hopefully I did that right :-), I haven't run/tested any of this so there's a good chance I introduce a bug or three.

As a separate comment, I'm a little confused why the identifier groups (like `[$\w\$\.]`) have `$` in them. I get we want `\$` to match literal dollar signs but I don't think we need `$`. You might want to remove that and see if our test still passes, if so I'd get rid of it just to simplify things.